src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs

```diff
@@ -392,26 +361,120 @@ private void LowerBasicSwitch(DecisionTree.ByValue byValue)
+                            break;
+                        }
+
+                    default:
+                        {
+                            // other types, such as float, double, and decimal, are not currently
+                            // handled by emit and must be lowered here.
+                            _loweredDecisionTree.Add(LowerNonprimitiveSwitch(switchStatement));
+                            break;
+                        }
+                }
+
                 LowerDecisionTree(byValue.Expression, byValue.Default);
             }
 
+            private BoundStatement LowerNonprimitiveSwitch(BoundSwitchStatement switchStatement)
+            {
+                // Here we handle "other" types, such as float, double, and decimal, by
+                // lowering the BoundSwitchStatement.
+                // We compare the constant values using value.Equals(input), using ordinary
+                // overload resolution. Note that we cannot and do not rely on switching
+                // on the hash code, as it may not be consistent with the behavior of Equals;
+                // see https://github.com/dotnet/coreclr/issues/6237. Also, the hash code is
+                // not guaranteed to be the same on the compilation platform as the runtime
+                // platform.
+                // CONSIDER: can we improve the quality of the code using comparisons, like
+                //           we do for other numeric types, by generating a series of tests
+                //           that use divide-and-conquer to efficiently find a matching value?
+                //           If so, we should use the BoundSwitchStatement and do that in emit.
+                //           Moreover, we should be able to use `==` rather than `.Equals`
+                //           for cases (such as non-NaN) where we know the result to be the same.
+                var rewrittenSections = switchStatement.SwitchSections;
+                var expression = switchStatement.Expression;
+                var noValueMatches = switchStatement.BreakLabel;
+                Debug.Assert(switchStatement.LoweredPreambleOpt == null);
+                Debug.Assert(switchStatement.InnerLocals.IsDefaultOrEmpty);
+                Debug.Assert(switchStatement.InnerLocalFunctions.IsDefaultOrEmpty);
+                Debug.Assert(switchStatement.StringEquality == null);
+                LabelSymbol nextLabel = null;
+                var builder = ArrayBuilder<BoundStatement>.GetInstance();
+                foreach (var section in rewrittenSections)
+                {
+                    foreach (var boundSwitchLabel in section.SwitchLabels)
+                    {
+                        if (nextLabel != null)
+                        {
+                            builder.Add(_factory.Label(nextLabel));
+                        }
+                        nextLabel = _factory.GenerateLabel("failcase+" + section.SwitchLabels[0].ConstantValueOpt.Value);
```

Comment:
As it turns out, there will only be one label in each section for the switch statements that ever get to this translation code. So it won't make any difference.

---
In reply to: [70733585](https://github.com/dotnet/roslyn/pull/12478#discussion_r70733585) [](ancestors = 70733585)